# Segmenting and Clustering Neighborhoods in Toronto

## Part 1

Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas  dataframe

In [81]:
import pandas as pd
import numpy as np

In [105]:
urlread=pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
toronto_df=urlread[0]

# Ignore cells with a borough that is Not assigned
toronto_df.drop(toronto_df[toronto_df['Borough']=='Not assigned'].index,inplace=True)
toronto_df=toronto_df.reset_index()
toronto_df.drop(columns=['index'],inplace=True)

# However, this wiki page seems to be cleaned up already and does not have duplicate issue with more than one neighborhood can exist in one postal code area
gptoronto_df=toronto_df.groupby(['Borough'])
# print(gptoronto_df.head(103))

# empty! cannot find 'Neighbourhood' with 'Not assigned' values
print(toronto_df[(toronto_df['Neighbourhood']=='Not assigned')])

toronto_df['Postal Code'].value_counts()

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []


M7Y    1
M8V    1
M4P    1
M4M    1
M4Y    1
      ..
M4J    1
M2P    1
M9R    1
M3B    1
M1S    1
Name: Postal Code, Length: 103, dtype: int64

Here is the dataframe object

## Part 1 results

In [108]:
toronto_df.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                        Neighbourhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5             Islington Avenue, Humber Valley Village  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

In [110]:
# print the shape
print(toronto_df.shape)

(103, 3)


## Part 2

Now that we have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood. 

First, I would like to test the code shared in the instruction (however, it does not work out though)


In [111]:
import geocoder # import geocoder

# initialize your variable to None
lat_lng_coords = None
postal_code = 'M5G'

# loop until you get the coordinates
while(lat_lng_coords is None):
    g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
    lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]
print(lat_lng_coords)

KeyboardInterrupt: 

However, I cannot manage to establish the connection and obtain the coordinate data from the server.
I tested multiple times, but always time out from waiting so had to terminate the code.

Now I am going to use the different method by loading directly the coordinate data provided in the instruction



In [112]:
loc_coordinate=pd.read_csv('Geospatial_Coordinates.csv')
loc_coordinate

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

## Part 2 results
Now we can merge the two dataframes and combine the information to store the latitude/longitude information for the neighborhood.


In [114]:
toronto_coor=pd.merge(toronto_df,loc_coordinate,on='Postal Code')
toronto_coor.shape
toronto_coor.head(12)


Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                        Neighbourhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

## Question 3

### Goal: Now with the neighborhood information for Toronto, I am going to explore the neighborhood information and cluster the neighbors.

### First, I would like to limit myself to only the neighborhood with name containing "Toronto" near the downtown area for the clustsering and analysis

In [119]:
import folium


df_toronto_only=toronto_coor[toronto_coor.Borough.str.contains('Toronto',case=False)]
df_toronto_only=df_toronto_only.reset_index()
df_toronto_only.drop(columns=['index'],inplace=True)
toronto_lat=df_toronto_only['Latitude'].mean()
toronto_lon=df_toronto_only['Longitude'].mean()

print(df_toronto_only.shape)
df_toronto_only.head()

(39, 5)


Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

### For these selected Toronto neighborhoods, I would like to show where they are on the map first

In [124]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[toronto_lat,toronto_lon], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto_only['Latitude'], df_toronto_only['Longitude'], df_toronto_only['Borough'], df_toronto_only['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Credentials to use Foursquare API service.

In [76]:
CLIENT_ID = 'LKOW2YB34S5K4ICEYAYJZZGYEFNTBXN2EWH2WUTDULLD3MJK' # your Foursquare ID
CLIENT_SECRET = 'ZWI3LAYL1EYF0XJH1TEAUCHRAA4LBOOYVE0CCDIYLRBW1KPI' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: LKOW2YB34S5K4ICEYAYJZZGYEFNTBXN2EWH2WUTDULLD3MJK
CLIENT_SECRET:ZWI3LAYL1EYF0XJH1TEAUCHRAA4LBOOYVE0CCDIYLRBW1KPI


### My test here to try to run the API by Foursquare and to fetch the data near the given coordinate

Adopted from the instruction/tutorial code on Coursera

In [126]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius


neighborhood_latitude=df_toronto_only.loc[0, 'Latitude']
neighborhood_longitude=df_toronto_only.loc[0, 'Longitude']

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
print(url)

import requests
results = requests.get(url).json()
results

https://api.foursquare.com/v2/venues/explore?&client_id=LKOW2YB34S5K4ICEYAYJZZGYEFNTBXN2EWH2WUTDULLD3MJK&client_secret=ZWI3LAYL1EYF0XJH1TEAUCHRAA4LBOOYVE0CCDIYLRBW1KPI&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100


{'meta': {'code': 200, 'requestId': '6022851d0ec9ad2950d1e8cd'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 45,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '53b8466a498e83df908c3f21',
       'name': 'Tandem Coffee',
       'location': {'address': '368 King St E',
        'crossStreet': 'at Trinity St',
        'lat': 43.65355870959944,
        'lng': -79.36180945913513,
        'labeledLatLngs': [{'label': 'display',
 

In [127]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [130]:
from pandas.io.json import json_normalize
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

print(nearby_venues.shape)
nearby_venues.head()

(45, 4)


/Users/shiangfang/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  after removing the cwd from sys.path.


name           categories        lat        lng
0           Tandem Coffee          Coffee Shop  43.653559 -79.361809
1        Roselle Desserts               Bakery  43.653447 -79.362017
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3     Body Blitz Spa East                  Spa  43.654735 -79.359874
4          Impact Kitchen           Restaurant  43.656369 -79.356980

## Explore Neighborhoods in Toronto

Adopted from the tutorial/coursera code
Here I will go throiugh the list of Toronto neighborhoods and fetch the local information for each neighborhood. The information explores the local venues around the location. Later we will use these to classify the neighborhood regions.

In [132]:
LIMIT = 100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [134]:
toronto_venues = getNearbyVenues(names=df_toronto_only['Neighbourhood'],
                                   latitudes=df_toronto_only['Latitude'],
                                   longitudes=df_toronto_only['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [135]:
print(toronto_venues.shape)
toronto_venues.head()

(1588, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0           Tandem Coffee       43.653559       -79.361809   
1        Roselle Desserts       43.653447       -79.362017   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0          Coffee Shop  
1               Bakery  
2  Distribution Center  
3                  Spa  
4           Restaurant

In [137]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            54   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     18   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     62   
Christie                                                               15   
Church and Wellesley                                                   77   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        8   
Dufferin, Dovercourt Village                                           16   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
India Bazaar, The Beaches West                                         21   
Kensington Market, Chinatown, Grange Park                              59   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               42   
Moore Park, Summerhill East                                             2   
North Toronto West, Lawrence Park                                      19   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            30   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede, Swansea                                                     36   
St. James Town                                                         79   
St. James Town, Cabbagetown                                            42   
Stn A PO Boxes                                                         97   
Studio District                                                        36   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    20   
The Beaches                                                             4   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         30   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             54   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      18   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      62   
Christie                                                                15   
Church and Wellesley                                                    77   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              35   
Davisville North                    

In [138]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 227 uniques categories.


Now I am going to use the one-hot encoding for the local venues types and countings. (total 227 types of them as shown above!)


In [144]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

print(toronto_onehot.shape)
toronto_onehot.head()



(1588, 227)


Yoga Studio  Adult Boutique  Airport  Airport Food Court  Airport Gate  \
0            0               0        0                   0             0   
1            0               0        0                   0             0   
2            0               0        0                   0             0   
3            0               0        0                   0             0   
4            0               0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Theme Restaurant  Tibetan Restaurant  Toy / Game Store  \
0             0  ...                 0                   0                 0   
1             0  ...                 0                   0                 0   
2             0  ...                 0                   0                 0   
3             0  ...                 0                   0                 0   
4             0  ...                 0                   0                 0   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0      0              0                              0                 0   
1      0              0                              0                 0   
2      0              0                              0                 0   
3      0              0                              0                 0   
4      0              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  
0                      0         0          0  
1                      0         0          0  
2                      0         0          0  
3                      0         0          0  
4                      0         0          0  

[5 rows x 227 columns]

In [145]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   Business reply mail Processing Centre, South C...     0.055556   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                  Central Bay Street     0.016129   
5                                            Christie     0.000000   
6                                Church and Wellesley     0.025974   
7                      Commerce Court, Victoria Hotel     0.000000   
8                                          Davisville     0.000000   
9                                    Davisville North     0.000000   
10                       Dufferin, Dovercourt Village     0.000000   
11             First Canadian Place, Underground city     0.000000   
12    Forest Hill North & West, Forest Hill Road Park     0.000000   
13                           Garden District, Ryerson     0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000   
15                      High Park, The Junction South     0.000000   
16                     India Bazaar, The Beaches West     0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000   
18                                      Lawrence Park     0.000000   
19                           Little Portugal, Trinity     0.023810   
20                        Moore Park, Summerhill East     0.000000   
21                  North Toronto West, Lawrence Park     0.052632   
22                             Parkdale, Roncesvalles     0.000000   
23        Queen's Park, Ontario Provincial Government     0.033333   
24                          Regent Park, Harbourfront     0.022222   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.027778   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                     Stn A PO Boxes     0.010309   
32                                    Studio District     0.027778   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                       The Danforth West, Riverdale     0.023810   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.033333   

    Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0         0.000000   0.0000              0.0000        0.0000           0.000   
1         0.000000   0.0000              0.0000        0.0000           0.000   
2         0.000000   0.0000              0.0000        0.0000           0.000   
3         0.000000   0.0625              0.0625        0.0625           0.125   
4         0.000000   0.0000              0.0000        0.0000           0.000   
5         0.000000   0.0000              0.0000        0.0000           0.000   
6         0.012987   0.0000              0.0000        0.0000           0.000   
7         0.000000   0.0000              0.0000        0.0000           0.000   
8         0.000000   0.0000              0.0000        0.0000           0.000   
9         0.000000   0.0000              0.0000        0.0000           0.000   
10        0.000000   0.0000              0.0000        0.0000           0.000   
11        0.000000   0.0000              0.0000        0.0000           0.000   
12        0.000000   0.0000              0.0000        0.0000           0.000   
13        0.000000   0.0000              0.0000        0.0000           0.000   
14        0.000000   0.00

In [147]:
toronto_grouped.shape

(39, 227)

### Here I show the top 5/10 (by counts/ratio) types of venues within each neighborhood region

In [149]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.06
2              Bakery  0.04
3  Seafood Restaurant  0.04
4         Cheese Shop  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1  Breakfast Spot  0.09
2          Bakery  0.09
3     Coffee Shop  0.09
4             Gym  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0    Light Rail Station  0.11
1           Yoga Studio  0.06
2         Auto Workshop  0.06
3            Smoke Shop  0.06
4  Fast Food Restaurant  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                 venue  freq
0      Airport Service  0.19
1       Airport Lounge  0.12
2     Airport Terminal  0.12
3             Boutique  0.06
4  Rental Car Location  0.06


----Central Bay Street----
                ve

In [150]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [153]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                Bakery    Seafood Restaurant   
1        Breakfast Spot                Bakery           Coffee Shop   
2           Yoga Studio         Auto Workshop            Smoke Shop   
3        Airport Lounge      Airport Terminal              Boutique   
4        Sandwich Place                  Café    Italian Restaurant   

  5th Most Common Venue   6th Most Common Venue 7th Most Common Venue  \
0           Cheese Shop              Restaurant        Farmers Market   
1                   Gym  Furniture / Home Store                   Bar   
2  Fast Food Restaurant          Farmers Market         Burrito Place   
3   Rental Car Location                 Airport    Airport Food Court   
4          Burger Joint             Salad Place       Bubble Tea Shop   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   Beer Bar             Jazz Club            Fish Market  
1                  Nightclub          Climbing Gym             Restaurant  
2                        Spa            Restaurant                 Garden  
3               Airport Gate       Harbor / Marina       Sculpture Garden  
4  Middle Eastern Restaurant                  Park                  Diner

### With the information above, now I am going to clusetr the neighborhood regions and see which ones are similar / dissimilar when compared with others.

I am going to test with 3 clusters

In [177]:
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [178]:
# add clustering labels

print(neighborhoods_venues_sorted.head())
print(df_toronto_only.head())


neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronto_only

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

   Cluster Labels                                       Neighborhood  \
0               0                                        Berczy Park   
1               0       Brockton, Parkdale Village, Exhibition Place   
2               0  Business reply mail Processing Centre, South C...   
3               0  CN Tower, King and Spadina, Railway Lands, Har...   
4               0                                 Central Bay Street   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop          Cocktail Bar                Bakery   
1                  Café        Breakfast Spot                Bakery   
2    Light Rail Station           Yoga Studio         Auto Workshop   
3       Airport Service        Airport Lounge      Airport Terminal   
4           Coffee Shop        Sandwich Place                  Café   

  4th Most Common Venue 5th Most Common Venue   6th Most Common Venue  \
0    Seafood Restaurant           Cheese Shop              Restaura

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0           Coffee Shop   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               0     Health Food Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Pub                Bakery                  Park   
1      Sushi Restaurant           Yoga Studio   Fried Chicken Joint   
2        Clothing Store                  Café   Japanese Restaurant   
3          Cocktail Bar                  Café   American Restaurant   
4                 Trail                   Pub           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0        Breakfast Spot               Theater                       Café   
1         Burrito Place                  Café         College Auditorium   
2        Cosmetics Shop    Italian Restaurant  Middle Eastern Restaurant   
3             Gastropub                   Gym                   Creperie   
4                Museum   Martial Arts School   Mediterranean Restaurant   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0           Event Space        Chocolate Shop             Cosmetics Shop  
1              Creperie                 Diner        Distribution Center  
2       Bubble Tea Shop     Electronics Store                Pizza Place  
3                Bakery        Clothing Store         Seafood Restaurant  
4           Men's Store    Mexican Restaurant  Middle Eastern Restaurant

### Use the map with colors labeled for neighborhood types/clusters

In [181]:
import matplotlib.cm as cm
import matplotlib.colors as colors
# create map
map_clusters = folium.Map(location=[toronto_lat,toronto_lon], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Let me give some more information about what neighborhood regions are included within each cluster

#### Cluster 1

In [165]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3   Downtown Toronto               0           Coffee Shop   
4       East Toronto               0     Health Food Store   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0         Grocery Store   
8   Downtown Toronto               0           Coffee Shop   
9       West Toronto               0              Pharmacy   
10  Downtown Toronto               0           Coffee Shop   
11      West Toronto               0                   Bar   
12      East Toronto               0      Greek Restaurant   
13  Downtown Toronto               0           Coffee Shop   
14      West Toronto               0                  Café   
15      East Toronto               0        Sandwich Place   
16  Downtown Toronto               0           Coffee Shop   
17      East Toronto               0           Coffee Shop   
20   Central Toronto               0  Gym / Fitness Center   
22      West Toronto               0    Mexican Restaurant   
23   Central Toronto               0        Clothing Store   
24   Central Toronto               0                  Café   
25      West Toronto               0             Gift Shop   
26   Central Toronto               0           Pizza Place   
27  Downtown Toronto               0                  Café   
28      West Toronto               0                  Café   
30  Downtown Toronto               0                  Café   
31   Central Toronto               0           Coffee Shop   
32  Downtown Toronto               0       Airport Service   
34  Downtown Toronto               0           Coffee Shop   
35  Downtown Toronto               0                  Café   
36  Downtown Toronto               0           Coffee Shop   
37  Downtown Toronto               0           Coffee Shop   
38      East Toronto               0    Light Rail Station   

            2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                             Pub                Bakery                  Park   
1                Sushi Restaurant           Yoga Studio   Fried Chicken Joint   
2                  Clothing Store                  Café   Japanese Restaurant   
3                    Cocktail Bar                  Café   American Restaurant   
4                           Trail                   Pub           Yoga Studio   
5                    Cocktail Bar                Bakery    Seafood Restaurant   
6                  Sandwich Place                  Café    Italian Restaurant   
7                            Café                  Park             Nightclub   
8                            Café            Restaurant                   Gym   
9                          Bakery             Pet Store                  Pool   
10                       Aquarium                 Hotel                  Café   
11  Vegetarian / Vegan Restaurant           Men's Store                  Café   
12                    Coffee Shop    Italian Restaurant        Ice Cream Shop   
13                          Hotel                  Café           Salad Place   
14                 Breakfast Spot                Bakery           Coffee Shop   
15           Fast Food Restaurant                   Gym      Sushi Restaurant   
16                     Restaurant                 Hotel                  Café   
17                         Bakery               Brewery   American Restaurant   
20                 Breakfast Spot      Department Store          Dance Studio   
22                           Café       Thai Restaurant   Arts & Crafts Store   
23                    Coffee Shop           Yoga Studio   Sporting Goods Shop   
24                 Sandwich Place           Coffee Shop     

#### Cluster 2

In [167]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
29   Central Toronto               1                  Park   
33  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29                 Trail           Yoga Studio         Movie Theater   
33            Playground                 Trail           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
29                Market   Martial Arts School  Mediterranean Restaurant   
33         Movie Theater                Market       Martial Arts School   

       8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
29               Men's Store    Mexican Restaurant  Middle Eastern Restaurant  
33  Mediterranean Restaurant           Men's Store         Mexican Restaurant

#### Cluster 3

In [168]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
21  Central Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21                 Trail         Jewelry Store      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21           Yoga Studio   Moroccan Restaurant                Market   

   8th Most Common Venue     9th Most Common Venue 10th Most Common Venue  
21   Martial Arts School  Mediterranean Restaurant            Men's Store

#### Cluster 4

In [169]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18           Swim School              Bus Line           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
18         Movie Theater   Martial Arts School  Mediterranean Restaurant   

   8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
18           Men's Store    Mexican Restaurant  Middle Eastern Restaurant

#### Cluster 5

In [171]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               4  Fast Food Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19          Home Service                Garden                Museum   

   5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
19                Market   Martial Arts School  Mediterranean Restaurant   

   8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
19           Men's Store    Mexican Restaurant  Middle Eastern Restaurant